### PR TITLE
solana-ibc: use wallet addresses instead of token accounts in packets

### PR DESCRIPTION
**Pull request by dhruvja**
Currently we were using token accounts as sender/receiver in the packets. But coming to think of it, its better to use wallet addresses since token accounts can be easily derived from the token mint and wallet address. It also makes it easy to send a transfer from other chain by just sending it to wallet address rather than deriving token account on client side and sending it since that cannot be trusted. Additionally we have checks to ensure that the token account to which the tokens are transferred to are always owned by the sender/receiver.

Also in `unescrow_validate` using `Unescrow` tag instead of `Escrow` tag for validation.

### CODE REVIEW

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 1]
```diff
@@ -287,10 +293,31 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
...
             .ok_or(TokenTransferError::ParseAccountFailure)?;
-        if !account.0.eq(token_account.key) {
+        let receiver = accounts
+            .receiver
+            .as_ref()
+            .ok_or(TokenTransferError::ParseAccountFailure)?;
+
+        let receiver_token_account = get_token_account(account.0, token_mint);
+
+        if !account.0.eq(receiver.key) {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:303]:
```suggestion
        if account.0 != *receiver.key {
```
and same throughout the PR.

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 2]
```diff
@@ -73,6 +73,12 @@ fn get_token_mint(denom: &PrefixedDenom) -> Result<Pubkey, TokenTransferError> {
     Ok(Pubkey::find_program_address(&seeds, &crate::ID).0)
 }
 
+fn get_token_account(owner: Pubkey, token_mint: Pubkey) -> Pubkey {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:76]:
```suggestion
fn get_token_account(owner: &Pubkey, token_mint: &Pubkey) -> Pubkey {
```

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 3]
```diff
@@ -388,17 +436,26 @@ impl IbcStorage<'_, '_> {
...
-        accounts
-            .token_account
-            .as_ref()
-            .filter(|token_account| account.0.eq(token_account.key))
-            .ok_or(TokenTransferError::ParseAccountFailure)?;
-
+        // We only need to check for sender/receiver since the token account
+        // is always derived from the token mint so if sender/receiver are right,
+        // the token account would be right as well.
         let ok = match op {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:442]: I’m not too happy with this match here.  On one side we have `ok_or`
with question mark operator.  On the other the `ok` variable is still
used.  Then again I don’t know if there’s some cleaner way to do it.
Perhaps something like:

```
        match op {
            EscrowOp::Escrow => {
                accounts
                    .sender
                    .as_deref()
                    .filter(|sender| sender.is_signer)
            },
            EscrowOp::Unescrow => {
                accounts
                    .receiver
                    .as_deref()
                    .filter(|_| accounts.mint_authority.is_some())
            },
        }
            .filter(|acc| account.0 == *acc.key)
            .map(|_| ())
            .ok_or(TokenTransferError::ParseAccountFailure)
```

**Review by mina86 — APPROVED** [review]

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs (hunk 2 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:76]: gotcha

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs (hunk 1 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:303]: right

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs (hunk 3 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:442]: i think above is good enough

File: solana/solana-ibc/programs/solana-ibc/src/tests.rs [hunk 4]
```diff
@@ -720,9 +720,9 @@ fn anchor_test_deliver() -> Result<()> {
         channel_id_on_b.clone(),
         channel_id_on_a.clone(),
         3,
-        sender_token_address,
-        receiver_native_token_address,
-        String::new(),
+        authority.pubkey(),
+        receiver.pubkey(),
+        String::from(""),
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/tests.rs:725]: But why not `String::new`?

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 5]
```diff
@@ -287,10 +293,31 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
...
+            .as_ref()
+            .ok_or(TokenTransferError::ParseAccountFailure)?;
+
+        let receiver_token_account = get_token_account(&account.0, &token_mint);
+
+        if account.0 != *receiver.key {
+            msg!("Token account not found {} {:?}", account, receiver.key);
             return Err(TokenTransferError::ParseAccountFailure);
         }
         if !token_mint.eq(token_mint_account.key) {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:307]:
```suggestion
        if token_mint != *token_mint_account.key {
```

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 6]
```diff
@@ -287,10 +293,31 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
...
+            msg!(
+                "Token mint not found {:?} {:?}",
+                token_mint,
+                token_mint_account.key
+            );
+            return Err(TokenTransferError::ParseAccountFailure);
+        }
+        if !receiver_token_account.eq(token_account.key) {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:315]:
```suggestion
        if receiver_token_account != *token_account.key {
```

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 7]
```diff
@@ -328,10 +355,31 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
...
+            .as_ref()
+            .ok_or(TokenTransferError::ParseAccountFailure)?;
+
+        let sender_token_account = get_token_account(&account.0, &token_mint);
+
+        if account.0 != *sender.key {
+            msg!("Token account not found {} {:?}", account, sender.key);
             return Err(TokenTransferError::ParseAccountFailure);
         }
         if !token_mint.eq(token_mint_account.key) {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:369]:
```suggestion
        if token_mint != *token_mint_account.key {
```

File: solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs [hunk 8]
```diff
@@ -328,10 +355,31 @@ impl TokenTransferValidationContext for IbcStorage<'_, '_> {
...
+            msg!(
+                "Token mint not found {:?} {:?}",
+                token_mint,
+                token_mint_account.key
+            );
+            return Err(TokenTransferError::ParseAccountFailure);
+        }
+        if !sender_token_account.eq(token_account.key) {
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs:377]:
```suggestion
        if sender_token_account != *token_account.key {
```

**Review by mina86 — APPROVED** [review]

File: solana/solana-ibc/programs/solana-ibc/src/tests.rs (hunk 4 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/tests.rs:725]: sure

**Pull request merged**